src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/AbstractImportCompletionProvider.cs

```diff
@@ -47,6 +48,8 @@ public override async Task ProvideCompletionsAsync(CompletionContext completionC
             var syntaxContext = await CreateContextAsync(document, completionContext.Position, cancellationToken).ConfigureAwait(false);
             if (!ShouldProvideCompletion(completionContext, syntaxContext))
             {
+                // fire and forget
+                _ = Task.Run(() => WarmUpCacheAsync(document));
```

Comment:
this worries me.  how many of these might be in flight at a time?  can you use an AsyncBatchingWorkQueue so we only ever are tyring to execute the warmup code at most one at a time?